src/metabase/models/query/permissions.clj

```diff
@@ -74,11 +74,17 @@
        (:source-table m)
        (query->source-table-ids (dissoc m :source-table)))))))
 
+(def ^:dynamic *card-instances*
+  "A map from card IDs to card instances with the collection_id (possibly nil).
+  Useful when bulk loading cards from different databases."
+  nil)
```

Comment:
This is a hacky ad-hoc solution, suggestions for better ones are welcome.

Reply:
I think a `with-card-instances` in this namespace would be cleaner. Make it  use and bind with your bulk select . Make `card-instance` use your bulk select too so that query-perms is the source of truth of what needs to be selected for an instance.